### PR TITLE
🐛Source Google ads: connector doesn't pull data for today

### LINKED ISSUE #7047: Source Google ads: connector doesn't pull data for today
## Current Behavior
Data is not pulled for today. https://github.com/airbytehq/airbyte/blob/3b32a091ff6e9fc502ffeb284e8550b05ca8f300/airbyte-integrations/connectors/source-google-ads/source_google_ads/streams.py#L88 

## Expected Behavior
It should also pull for the same date

## Logs
*If applicable, please upload the logs from the failing operation. 
For sync jobs, you can download the full logs from the UI by going to the sync attempt page and 
clicking the download logs button at the top right of the logs display window.*

<details>
<summary>LOG</summary> 

```

replace this with
your long log
output here

```

</details>

## Steps to Reproduce
1.
2.
3.

## Are you willing to submit a PR?

Remove this with your answer.

### PR DESCRIPTION
**Pull request by annalvova05**
## What
For incremental streams `end_date` param is yesterday's date now. Data is not pulled for today. Could reproduce it locally.

Closes #7047 and OnCall[#37](https://github.com/airbytehq/oncall/issues/37).

## How
Update date to today and take timezone into account.

## Recommended reading order
1. `streams.py`

## 🚨 User Impact 🚨
Are there any breaking changes? If yes, please make sure to include it here and in any changelogs with the 🚨🚨 emoji
What is the end result perceived by the user?

## Pre-merge Checklist
Expand the relevant checklist and delete the others. 

<details><summary> <strong> Updating a connector </strong></summary>
<p>
   
#### Community member or Airbyter
   
- [ ] Grant edit access to maintainers ([instructions](https://docs.github.com/en/github/collaborating-with-pull-requests/working-with-forks/allowing-changes-to-a-pull-request-branch-created-from-a-fork#enabling-repository-maintainer-permissions-on-existing-pull-requests))
- [ ] Secrets in the connector's spec are annotated with `airbyte_secret` 
- [ ] Unit & integration tests added and passing. Community members, please provide proof of success locally e.g: screenshot or copy-paste unit, integration, and acceptance test output. To run acceptance tests for a Python connector, follow instructions in the README. For java connectors run `./gradlew :airbyte-integrations:connectors:<name>:integrationTest`.
- [ ] Code reviews completed
- [ ] Documentation updated 
    - [ ] Connector's `README.md`
    - [ ] Connector's `bootstrap.md`. See [description and examples](https://docs.google.com/document/d/1ypdgmwmEHWv-TrO4_YOQ7pAJGVrMp5BOkEVh831N260/edit?usp=sharing)
    - [ ] Changelog updated in `docs/integrations/<source or destination>/<name>.md` including changelog. See changelog [example](https://docs.airbyte.io/integrations/sources/stripe#changelog)
- [ ] PR name follows [PR naming conventions](https://docs.airbyte.io/contributing-to-airbyte/updating-documentation#issues-and-pull-requests)
   
#### Airbyter

If this is a community PR, the Airbyte engineer reviewing this PR is responsible for the below items. 
   
- [ ] Create a non-forked branch based on this PR and test the below items on it
- [ ] Build is successful
- [ ] Credentials added to Github CI. [Instructions](https://docs.airbyte.io/connector-development#using-credentials-in-ci). 
- [ ] [`/test connector=connectors/<name>` command](https://docs.airbyte.io/connector-development#updating-an-existing-connector) is passing. 
- [ ] New Connector version released on Dockerhub by running the `/publish` command described [here](https://docs.airbyte.io/connector-development#updating-an-existing-connector)
- [ ] After the new connector version is published, connector version bumped in the seed directory as described [here](https://docs.airbyte.io/connector-development#publishing-a-connector)
- [ ] Seed specs have been re-generated by building the platform and committing the changes to the seed spec files, as described [here](https://docs.airbyte.io/connector-development#publishing-a-connector)

</p>
</details>

### CODE REVIEW

File: airbyte-integrations/connectors/source-google-ads/source_google_ads/streams.py [hunk 1]
```diff
@@ -71,6 +72,14 @@ def __init__(self, start_date: str, conversion_window_days: int, **kwargs):
...
         super().__init__(**kwargs)
 
+    @property
+    def time_zone(self):
+        account = next(Accounts(self.google_ads_client).read_records(sync_mode=None), {})
+        time_zone_name = account.get("customer.time_zone")
+        if time_zone_name:
+            return Timezone(time_zone_name)
+        return "local"
+
```
Comment (antixar) [inline airbyte-integrations/connectors/source-google-ads/source_google_ads/streams.py:82]: Please add an annotation for returnable values. Can this function return one type any time?

Comment (antixar) [reply airbyte-integrations/connectors/source-google-ads/source_google_ads/streams.py:82]: the property "time_zone" can be called several times during one sync. Please add some cache mechanism for its value

**Review by antixar — CHANGES_REQUESTED** [review]
Please add an unit test for this case when a time_zone value can influence data-time values.

File: airbyte-integrations/connectors/source-google-ads/source_google_ads/streams.py [hunk 2]
```diff
@@ -83,9 +92,8 @@ def stream_slices(self, stream_state: Mapping[str, Any] = None, **kwargs) -> Ite
             days_of_data_storage=self.days_of_data_storage,
         )
 
-    @staticmethod
-    def get_date_params(stream_slice: Mapping[str, Any], cursor_field: str, end_date: pendulum.datetime = None, time_unit: str = "months"):
-        end_date = end_date or pendulum.yesterday()
```
Comment (midavadim) [inline airbyte-integrations/connectors/source-google-ads/source_google_ads/streams.py:88]: I think there were some reasons to do it.

I suspect that Google Ads do not generate reports in real time. 

As far as I remember google need some time (a day) to have full report for given day.

If we set end_day to today then report can be empty or incomplete. So please do additional research for this question

**Review by midavadim — COMMENTED** [review]
Please do not forget to update version and docs

File: airbyte-integrations/connectors/source-google-ads/source_google_ads/streams.py (hunk 2 above)
Comment (annalvova05) [reply airbyte-integrations/connectors/source-google-ads/source_google_ads/streams.py:88]: I didn't find any information about impossibility sync reports in real-time. Google Ads returns data for today

File: airbyte-integrations/connectors/source-google-ads/source_google_ads/spec.json [hunk 3]
```diff
@@ -52,7 +52,7 @@
       "customer_id": {
         "title": "Customer Id",
         "type": "string",
-        "description": "Customer id must be specified as a 10-digit number without dashes. More instruction on how to find this value in our <a href=\"https://docs.airbyte.io/integrations/sources/google-adwords#setup-guide\">docs</a>",
```
Comment (antixar) [inline airbyte-integrations/connectors/source-google-ads/source_google_ads/spec.json:55]: title = "Customer ID"

**Review by antixar — COMMENTED** [review]
Please add a minimal unit test for checking of available streams if the option login_customer_id is set or not.

File: airbyte-integrations/connectors/source-google-ads/source_google_ads/spec.json (hunk 3 above)
Comment (annalvova05) [reply airbyte-integrations/connectors/source-google-ads/source_google_ads/spec.json:55]: done. Added unit_tests

File: airbyte-integrations/connectors/source-google-ads/source_google_ads/streams.py (hunk 1 above)
Comment (annalvova05) [reply airbyte-integrations/connectors/source-google-ads/source_google_ads/streams.py]: updated, added unit_test

**Review by antixar — APPROVED** [review]
LGTM!

File: airbyte-integrations/connectors/source-google-ads/source_google_ads/source.py [hunk 4]
```diff
@@ -56,24 +67,37 @@ def check_connection(self, logger: AirbyteLogger, config: Mapping[str, Any]) ->
...
-        ] + custom_query_streams
+        ]
+
+        custom_query_streams = [
+            CustomQuery(custom_query_config=single_query_config, **incremental_stream_config)
+            for single_query_config in config.get("custom_queries", [])
+        ]
+        streams.extend(custom_query_streams)
+
+        is_manager_account = "login_customer_id" in config and config["login_customer_id"].strip()
```
Comment (sherifnada) [inline airbyte-integrations/connectors/source-google-ads/source_google_ads/source.py:92]: I think this is not foolproof. The contract is: 

`customer_id` could be either manager account or non-manager account. If it's non-manager, then `login_customer_id` should be set to the ID of the manager account. So just looking at the config might not be sufficient

Comment (annalvova05) [reply airbyte-integrations/connectors/source-google-ads/source_google_ads/source.py:92]: If I understand correctly, there may be two cases:
1. When [customer_id](https://developers.google.com/google-ads/api/docs/first-call/overview?hl=en) is specified as manager account. In this case, we can get data from all streams.
2. When `customer_id` is specified as a non-manager account and `login_customer_id` is [specified](https://developers.google.com/google-ads/api/docs/concepts/call-structure#cid) as a manager account. In this case, not all streams are available.

Or are you talking about adding a check that the `customer_id` is a manager account, then making available all streams?

File: airbyte-integrations/connectors/source-google-ads/source_google_ads/streams.py (hunk 2 above)
Comment (midavadim) [reply airbyte-integrations/connectors/source-google-ads/source_google_ads/streams.py:88]: @sherifnada We need your opinion here.

1. As for me if we include 'today'  into the current sync means that next sync starts from tomorrow so client will lose data between now and the end of today's day. 

2. as far as I remember from my own experience with Google Ads - it does not provide 'real-time' reports and full data is available only in a few hours .

If we want to include todays report into sync then an alternative solution can be - we should have 1 day overlaps between scans. It guarantees that customers have full reports.

Comment (midavadim) [reply airbyte-integrations/connectors/source-google-ads/source_google_ads/streams.py:88]: Actually, I have found Google docs about it

https://support.google.com/google-ads/answer/2544985?hl=en

- such as clicks, conversions, and impressions) are delayed by less than 3 hours
- conversions attributed using attribution models other than “Last click” are typically delayed up to 15 hours
etc

File: airbyte-integrations/connectors/source-google-ads/source_google_ads/source.py (hunk 4 above)
Comment (sherifnada) [reply airbyte-integrations/connectors/source-google-ads/source_google_ads/source.py:92]: 1. when customer ID is a manager account, we cannot get data from all streams, only the ones available for manager account 
2. when customer ID is not a manager account, can get from all streams

Comment (annalvova05) [reply airbyte-integrations/connectors/source-google-ads/source_google_ads/source.py]: Updated. We request account information from Google Analytics and use the value from `customer.manager` field.

File: airbyte-integrations/connectors/source-google-ads/unit_tests/test_source.py [hunk 5]
```diff
@@ -17,10 +18,39 @@ def test_chunk_date_range():
...
+
+
+def test_non_manager_account():
+    mock_account_info = {"customer.manager": False}
+    source = SourceGoogleAds()
+    is_manager_account = source.is_manager_account(mock_account_info)
+    assert not is_manager_account
+
+
+def test_manager_account():
```
Comment (sherifnada) [inline airbyte-integrations/connectors/source-google-ads/unit_tests/test_source.py:35]: nice

File: docs/integrations/sources/google-ads.md [hunk 6]
```diff
@@ -35,6 +35,8 @@ This source is capable of syncing the following tables and their data:
 
 **Note**: Due to constraints from the Google Ads API, the `click_view` stream retrieves data one day at a time and can only retrieve data newer than 90 days ago
 
+**Note**: Due to constraints from the Google Ads API, [metrics](https://developers.google.com/google-ads/api/fields/v8/metrics) cannot be requested for a manager account. If we try to sync streams with metrics for the manager account, we will get the `REQUESTED_METRICS_FOR_MANAGER` [error](https://developers.google.com/google-ads/api/reference/rpc/v8/QueryErrorEnum.QueryError?hl=en#requested_metrics_for_manager). Metrics present in all report tables and all report streams. All report streams are unavailable for a manager account.
```
Comment (sherifnada) [inline docs/integrations/sources/google-ads.md:38]: good job adding this here! reducing impl details for user
```suggestion
**Note**: Due to constraints from the Google Ads API, [metrics](https://developers.google.com/google-ads/api/fields/v8/metrics) cannot be requested for a manager account. Therefore, report streams are only available when pulling data from a non-manager account. 
```

File: airbyte-integrations/connectors/source-google-ads/source_google_ads/streams.py (hunk 2 above)
Comment (sherifnada) [reply airbyte-integrations/connectors/source-google-ads/source_google_ads/streams.py:88]: I think it's acceptable to sync data only up to yesterday if there is no other way to guarantee data loss. Ideally, we can pick up a sync exactly where we left off. But if it's not possible, that's acceptable imo. 

What is the technical possibility of doing it that way?

File: airbyte-integrations/connectors/source-google-ads/source_google_ads/source.py [hunk 7]
```diff
@@ -56,24 +74,37 @@ def check_connection(self, logger: AirbyteLogger, config: Mapping[str, Any]) ->
...
+        streams = [
             AdGroupAds(api=google_api),
             AdGroups(api=google_api),
             Accounts(api=google_api),
             Campaigns(api=google_api),
             ClickView(**incremental_stream_config),
-        ] + custom_query_streams
+        ]
+
+        custom_query_streams = [
```
Comment (sherifnada) [inline airbyte-integrations/connectors/source-google-ads/source_google_ads/source.py:94]: a user could configure a custom query involving metrics with a manager account. Is it possible to check this here too, and exclude custom query streams from manager accounts if they contain metrics?

Comment (annalvova05) [reply airbyte-integrations/connectors/source-google-ads/source_google_ads/source.py:109]: add `is_metrics_in_custom_query` in check method

File: airbyte-integrations/connectors/source-google-ads/source_google_ads/streams.py (hunk 2 above)
Comment (annalvova05) [reply airbyte-integrations/connectors/source-google-ads/source_google_ads/streams.py:88]: @sherifnada we cannot sync data up to today. We use `segments.date` field for filtering data in incremental sync. Google Ads [accepts](https://developers.google.com/google-ads/api/fields/v8/ad_group_ad#segments.date) only date without time for filtering.

Comment (sherifnada) [reply airbyte-integrations/connectors/source-google-ads/source_google_ads/streams.py:88]: Ok, lets sync only until yesterday then. we can update that ticket with this information, and mention we can't do it due to limit from api side

File: airbyte-integrations/connectors/source-google-ads/source_google_ads/streams.py [hunk 8]
```diff
@@ -83,9 +92,8 @@ def stream_slices(self, stream_state: Mapping[str, Any] = None, **kwargs) -> Ite
...
+    def get_date_params(self, stream_slice: Mapping[str, Any], cursor_field: str, end_date: pendulum.datetime = None, time_unit: str = "months"):
+        end_date = end_date or pendulum.today(tz=self.time_zone)
```
Comment (sherifnada) [inline airbyte-integrations/connectors/source-google-ads/source_google_ads/streams.py]: shouldn't this be yesterday based on the comment above?

File: docs/integrations/sources/google-ads.md [hunk 9]
```diff
@@ -35,6 +35,10 @@ This source is capable of syncing the following tables and their data:
 
 **Note**: Due to constraints from the Google Ads API, the `click_view` stream retrieves data one day at a time and can only retrieve data newer than 90 days ago
 
+**Note**: Due to constraints from the Google Ads API, [metrics](https://developers.google.com/google-ads/api/fields/v8/metrics) cannot be requested for a manager account. Therefore, report streams are only available when pulling data from a non-manager account. 
+
+**Note**: For incremental streams we sync data up to yesterday using your Google Ads account time zone. Google Ads can filter data only by [date](https://developers.google.com/google-ads/api/fields/v8/ad_group_ad#segments.date) without time. Also, some report cannot load data in real time due to Google Ads [limitations](https://support.google.com/google-ads/answer/2544985?hl=en).
```
Comment (sherifnada) [inline docs/integrations/sources/google-ads.md:40]:
```suggestion
**Note**: For incremental streams data is synced up to the previous day using your Google Ads account time zone. The reason is that Google Ads can filter data only by [date](https://developers.google.com/google-ads/api/fields/v8/ad_group_ad#segments.date) without time. Also, some report cannot load data in real time due to Google Ads [limitations](https://support.google.com/google-ads/answer/2544985?hl=en).
```

**Review by sherifnada — APPROVED** [review]

File: airbyte-integrations/connectors/source-google-ads/source_google_ads/streams.py (hunk 8 above)
Comment (annalvova05) [reply airbyte-integrations/connectors/source-google-ads/source_google_ads/streams.py]: Yep, I returned it to yesterday.

**Pull request merged**